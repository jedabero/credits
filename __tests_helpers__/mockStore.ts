// eslint-disable-next-line import/no-extraneous-dependencies
import configureMockStore from 'redux-mock-store';
import { RootState } from '../src/store';

export default configureMockStore<RootState>();
