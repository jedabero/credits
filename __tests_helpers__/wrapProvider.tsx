import React, { ReactElement } from 'react';
import { Provider } from 'react-redux';
import { Store } from '../src/store';

const wrapProvider = (component: ReactElement, store: Store) => (
  <Provider store={store}>{component}</Provider>
);

export default wrapProvider;
