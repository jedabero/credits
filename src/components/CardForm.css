.cardrow > input {
  display: block;
  min-width: 0;
}
