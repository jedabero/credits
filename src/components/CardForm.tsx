import {
  useState, useCallback, useMemo, useEffect,
} from 'react';
import { EntityId } from '@reduxjs/toolkit';
import { selectCreditCardById, addCard, updateCard } from '../reducers/creditCards';
import { useAppSelector, useAppDispatch } from '../hooks';
import { CreditCard } from '../models/CreditCard';

type CardFormProps = {
  cardId?: EntityId,
  clearCardId: () => void,
};

function CardForm({ cardId, clearCardId }: CardFormProps) {
  const dispatch = useAppDispatch();
  const [wasSaved, setWasSaved] = useState(false);
  const id = useMemo(() => {
    if (wasSaved) setWasSaved(false);
    return cardId ?? crypto.randomUUID();
  }, [cardId, wasSaved]);
  const card = useAppSelector((state) => selectCreditCardById(state, id));
  const [values, setValues] = useState<CreditCard>({
    id,
    name: card?.name ?? '',
    cardNumber: card?.cardNumber ?? '',
    quota: card?.quota ?? 0,
    balance: card?.balance ?? 0,
    cutoffDay: card?.cutoffDay ?? 1,
  });
  useEffect(() => {
    setValues({
      id,
      name: card?.name ?? '',
      cardNumber: card?.cardNumber ?? '',
      quota: card?.quota ?? 0,
      balance: card?.balance ?? 0,
      cutoffDay: card?.cutoffDay ?? 1,
    });
  }, [id, card]);
  const setValue = useCallback(
    (event: React.ChangeEvent<HTMLInputElement & { name: keyof CreditCard }>) => {
      const { name, value } = event.target;
      const valueType = typeof values[name];
      if (valueType === 'number' && Number.isNaN(parseInt(value, 10))) return;
      setValues((oldValues: CreditCard) => ({
        ...oldValues,
        [name]: valueType === 'number' ? parseInt(value, 10) : value,
      }));
    },
    [values, setValues],
  );
  const save = useCallback(() => {
    if (cardId) {
      dispatch(updateCard({ id: cardId, changes: values }));
      clearCardId();
    } else {
      dispatch(addCard(values));
    }
    setWasSaved(true);
  }, [cardId, dispatch, clearCardId, values]);
  const cancel = useCallback(() => {
    setValues({
      id,
      name: '',
      cardNumber: '',
      quota: 0,
      balance: 0,
      cutoffDay: 1,
    });
    clearCardId();
  }, [id, clearCardId]);
  return (
    <div className="cardrow">
      <div>{(`${values.id}`).substring(0, 13)}</div>
      <input name="name" value={values.name} onChange={setValue} />
      <input name="cardNumber" value={values.cardNumber} onChange={setValue} />
      <input name="quota" type="number" value={values.quota} onChange={setValue} />
      <input name="balance" type="number" value={values.balance} onChange={setValue} />
      <input name="cutoffDay" type="number" value={values.cutoffDay} onChange={setValue} min={1} max={31} />
      <div>
        <button type="button" onClick={save}>{card?.id ? 'Update' : 'Create'}</button>
        {card?.id ? <button type="button" onClick={cancel}>Cancel</button> : null}
      </div>
    </div>
  );
}

CardForm.defaultProps = {
  cardId: null,
};

export default CardForm;
