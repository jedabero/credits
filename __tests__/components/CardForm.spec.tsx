/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { create } from 'react-test-renderer';
import CardForm from '../../src/components/CardForm';
import wrapProvider from '../../__tests_helpers__/wrapProvider';
import mockStore from '../../__tests_helpers__/mockStore';

describe('CardForm test', () => {
  test('should render', () => {
    const store = mockStore({
      _persist: { version: 0, rehydrated: true },
      creditCards: {
        entities: {},
        ids: [],
      },
      movements: {
        entities: {},
        ids: [],
      },
    });
    const tree = create(
      wrapProvider(
        <CardForm cardId="SOME ID" clearCardId={() => {}} />,
        store,
      ),
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
