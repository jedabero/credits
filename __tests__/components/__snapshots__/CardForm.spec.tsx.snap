// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardForm test should render 1`] = `
<div
  className="cardrow"
>
  <div>
    SOME ID
  </div>
  <input
    name="name"
    onChange={[Function]}
    value=""
  />
  <input
    name="cardNumber"
    onChange={[Function]}
    value=""
  />
  <input
    name="quota"
    onChange={[Function]}
    type="number"
    value={0}
  />
  <input
    name="balance"
    onChange={[Function]}
    type="number"
    value={0}
  />
  <input
    max={31}
    min={1}
    name="cutoffDay"
    onChange={[Function]}
    type="number"
    value={1}
  />
  <div>
    <button
      onClick={[Function]}
      type="button"
    >
      Create
    </button>
  </div>
</div>
`;
